components/sync15/src/coll_state.rs

```diff
@@ -0,0 +1,324 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use crate::error;
+use crate::key_bundle::KeyBundle;
+use crate::request::InfoConfiguration;
+use crate::state::GlobalState;
+use crate::sync::Store;
+use crate::util::ServerTimestamp;
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct CollSyncIds {
+    pub global: String,
+    pub coll: String,
+}
+
+/// Holds state for a collection. In general, only the CollState is
+/// needed to sync a collection (but a valid GlobalState is needed to obtain
+/// a CollState)
+#[derive(Debug, Clone)]
+pub struct CollState {
+    pub config: InfoConfiguration,
+    // initially from meta/global, updated after an xius POST/PUT.
+    pub last_modified: ServerTimestamp,
+    pub key: KeyBundle,
+}
+
+#[derive(Debug)]
+pub enum LocalCollState {
+    /// The state is unknown. If the collection knows what the IDs were last
+    /// time they are here, otherwise None
+    Unknown { ids: Option<CollSyncIds> },
+
+    /// The engine has been declined. This is a "terminal" state.
+    Declined,
+
+    /// There's no such collection in meta/global. We could possibly update
+    /// meta/global, but currently all known collections are there by default,
+    /// so this is, basically, an error condition.
+    NoSuchCollection,
+
+    /// Either the global or collection sync ID has changed - we will reset the engine.
+    SyncIdChanged { ids: CollSyncIds },
+
+    /// The collection is ready to sync.
+    Ready { key: KeyBundle },
+}
+
+pub struct LocalCollStateMachine<'state> {
+    global_state: &'state GlobalState,
+}
+
+impl<'state> LocalCollStateMachine<'state> {
+    fn advance(&self, from: LocalCollState, store: &Store) -> error::Result<LocalCollState> {
+        let name = &store.collection_name().to_string();
+        let meta_global = &self.global_state.global;
+        match from {
+            LocalCollState::Unknown { ids } => {
+                if meta_global.declined.contains(name) {
+                    return Ok(LocalCollState::Declined);
+                }
+                match meta_global.engines.get(name) {
```

Comment:
I feel like this match can be removed (e.g. reducing the nesting) if the right methods on Option are used. Could be wrong.

Reply:
This all got refactored with `StoreSyncAssoc`